alg/teca_regional_moisture_flux.cxx

```diff
@@ -0,0 +1,690 @@
+    {
+        TECA_ERROR("Failed to partition for the coordinate transform")
+        return -1;
+    }
+
+    // decompose for the divergence calc
+    unsigned long data_nelem = nlon*nlat;
+    dim3 data_tgrid;
+    dim3 data_bgrid;
+    int data_nblk = 0;
+    if (teca_cuda_util::partition_thread_blocks(device_id,
+        data_nelem, 8, data_bgrid, data_nblk, data_tgrid))
+    {
+        TECA_ERROR("Failed to partition for the divergence calculation")
+        return -1;
+    }
+
+    NESTED_VARIANT_ARRAY_DISPATCH_FP(lon.get(), _COORDS,
+        assert_type<TT_COORDS>(lat);
+
+        // transform the cooridnates
+        auto [splon, plon] = get_cuda_accessible<CTT_COORDS>(lon);
+        auto [splat, plat] = get_cuda_accessible<CTT_COORDS>(lat);
+
+        auto [theta, ptheta] = New<TT_COORDS>(nlon, allocator::cuda_async);
+        auto [phi, pphi] = New<CTT_COORDS>(nlat, allocator::cuda_async);
+        auto [sin_phi, psin_phi] = New<TT_COORDS>(nlat, allocator::cuda_async);
+
+        transform_coordinates<<<coord_bgrid,coord_tgrid>>>(ptheta,
+             pphi, psin_phi, plon, plat, nlon, nlat);
+
+        if ((ierr = cudaGetLastError()) != cudaSuccess)
+        {
+            TECA_ERROR("Failed to launch the transform_coordinates CUDA kernel"
+                << cudaGetErrorString(ierr))
+            return -1;
+        }
+
+        NESTED_VARIANT_ARRAY_DISPATCH_FP(ivt_u.get(), _DATA,
+            assert_type<TT_DATA>(ivt_v);
+
+            auto [spivt_u, pivt_u] = get_cuda_accessible<CTT_DATA>(ivt_u);
+            auto [spivt_v, pivt_v] = get_cuda_accessible<CTT_DATA>(ivt_v);
+
+            auto [div, pdiv] = New<TT_DATA>(data_nelem, NT_DATA(), allocator::cuda_async);
+            auto [cint, pcint] = New<TT_DATA>(data_nelem, NT_DATA(), allocator::cuda_async);
+
+            // divergence
+            divergence<<<data_bgrid, data_tgrid>>>(pdiv,
+                pivt_u, pivt_v, ptheta, pphi, psin_phi, nlon, nlat);
+
+            if ((ierr = cudaGetLastError()) != cudaSuccess)
+            {
+                TECA_ERROR("Failed to launch the divergence CUDA kernel"
+                    << cudaGetErrorString(ierr))
+                return -1;
+            }
+
+            // integrate cell by cell
+            integrate<<<data_bgrid, data_tgrid>>>(pcint, pdiv, ptheta, pphi, nlon, nlat);
+
+            if ((ierr = cudaGetLastError()) != cudaSuccess)
+            {
+                TECA_ERROR("Failed to launch the integrate CUDA kernel"
+                    << cudaGetErrorString(ierr))
+                return -1;
+            }
+
+            NESTED_VARIANT_ARRAY_DISPATCH(region_mask.get(), _MASK,
+
+                auto [spreg_mask, preg_mask] = get_cuda_accessible<CTT_MASK>(region_mask);
+
+                // sum over the masked region
+                moisture_flux = sum_if(pcint, preg_mask, data_nelem);
+
+                NT_DATA Re = 6378100.;
+                moisture_flux *= -Re;
```

Comment:
I made two comments above about a multiplication by (1/(Re*sin_phi[j]).  I see here that it looks like this is being taken care of here, which might be worth commenting on at the above locations.  I assume its to avoid multiplying everything by Re in the inner loop and instead doing it outside the loop?  But there's still the question of the 1/sin_phi -- maybe you did some transformation to avoid that term?

Reply:
> I made two comments above about a multiplication by (1/(Re*sin_phi[j]). I see here that it looks like this is being taken care of here, which might be worth commenting on at the above locations. I assume its to avoid multiplying everything by Re in the inner loop and instead doing it outside the loop? But there's still the question of the 1/sin_phi -- maybe you did some transformation to avoid that term?

Yes, `/(r_earth*sin(phi)` from the divergence cancels with the `r_earth^2*sin(phi)` from the dA. After a factor of `r_earth` remains. I pulled it out of the integral and apply it here, the very last step of  the calc. This all comes from eqn 8 in the overleaf, I will add this to the doxygen in the header file.

I can see how this is confusing the way the calculation is split up here into multiple steps. I split things up this way due to load balancing concerns on the GPU.

Reply:
makes sense, thanks!